honeybadgermpc/router.py

```diff
@@ -1,35 +1,67 @@
 import asyncio
-# import logging
+from abc import ABC, abstractmethod
+from functools import partial
+import logging
 
 
-def simple_router(n):
+class Router(ABC):
+    def __init__(self, num_parties, debug=False):
+        self.n = num_parties
+        self.sends = self._make_sends()
+        self.recvs = self._make_recvs()
+        self.broadcasts = self._make_broadcasts()
+        self.debug = debug
+
     """
-    Builds a set of connected channels
-    @return (receives, sends)
+    Base class for Router objects defining the recv and send methods.
     """
-    # Create a mailbox for each party
-    mbox = [asyncio.Queue() for _ in range(n)]
-
-    def make_send(i):
-        def _send(j, o):
-            # logging.debug('SEND %8s [%2d -> %2d]' % (o, i, j))
-            # delay = random.random() * 1.0
-            # asyncio.get_event_loop().call_later(delay, mbox[j].put_nowait,(i,o))
-            mbox[j].put_nowait((i, o))
-
-        return _send
-
-    def make_recv(j):
-        async def _recv():
-            (i, o) = await mbox[j].get()
-            # logging.debug('RECV %8s [%2d -> %2d]' % (o, i, j))
-            return (i, o)
-
-        return _recv
-
-    sends = {}
-    receives = {}
-    for i in range(n):
-        sends[i] = make_send(i)
-        receives[i] = make_recv(i)
-    return (sends, receives)
+    @abstractmethod
+    async def _recv(self, player_id: int) -> object:
+        """ Receives a message from player with id `player_id`
+        """
+        return NotImplementedError
+
+    @abstractmethod
+    def _send(self, player_id: int, dest_id: int, message: object):
+        """ Sends a message to player with id `dest` from `player_id`
+        """
+        return NotImplementedError
+
+    def _broadcast(self, player_id: int, message: object):
+        for dest_id in range(self.n):
+            self._send(player_id, dest_id, message)
+
+    def _make_recvs(self):
+        return [partial(self._recv, player_id) for player_id in range(self.n)]
+
+    def _make_sends(self):
+        return [partial(self._send, player_id) for player_id in range(self.n)]
+
+    def _make_broadcasts(self):
+        return [partial(self._broadcast, player_id) for player_id in range(self.n)]
+
+
+class SimpleRouter(Router):
+    def __init__(self, num_parties):
+        super().__init__(num_parties)
+
+        # Mailboxes for each party
+        self._queues = [asyncio.Queue() for _ in range(num_parties)]
+
+    async def _recv(self, player_id: int) -> object:
```

Comment:
why the `_` prefix? These are supposed to be public methods.

Reply:
At the time, the intention was to access these via the `.recvs` and `.sends` attributes, but I agree now.